### ISSUE TITLE
Step-CA shall act as a Registration Authority

### ISSUE THREAD

**Issue by dharanikumar-s** [smallstep/certificates#343]
### What would you like to be added
Step-CA shall act as a Registration Authority. Clients requests step-ca for signing or renewing a certificate, after authenticating and authorizing the client, step-ca will forward the request to external CA to get the certificate signed or renewed.

### Why this is needed
It is common among PKI implementation to function only as an registration authority and forwards the sign/renew requests to an external ca servers. This is required due to administrative/business reasons where all signing/renewing operations has to be done by an existing CA server, but that CA doesn't provide certificate provisioning APIs or present-day authentication or authorization framework for clients.

**Comment by maraino** [comment]
Hi @dharanikumar-s this is something that we've recently been thinking about, our KVM interface that we currently use to sign certificates using external HSMs currently does not support external CAs. But is on our plans to extend or use a similar interface to support external CAs.

@dharanikumar-s Is there any commercial external CA you have in mind? 

@rmhrisk, who reacted to this, is probably gonna be involved in one of the first integrations :) (I'm probably doing the actual  implementations, do we have docs 😇)

**Comment by hongkongkiwi** [comment]
I would love to use Hashicorp Vault for the actual signing request :-) And have smallstep in front handling ACME.

**Comment by mmalone** [comment]
@hongkongkiwi I'm curious: what's the advantage for you of having signing requests sent to Vault?

**Comment by mmalone** [comment]
@delamart thanks so much for the feedback!

FYI, if you're worried about keys in a folder, we do have pluggable key management / HSM backend support now. So you can keep your signing keys somewhere more secure. We have [docs for configuring plugins](https://github.com/smallstep/certificates/blob/master/docs/kms.md) for GCP CloudKMS, AWS KMS, and Yubikey. It's pretty easy to add a plugin if you want something different (the code lives [here](https://github.com/smallstep/certificates/tree/master/kms)).

None of this negates what you said. I think a Vault integration still makes sense. In fact, a Vault integration would probably start out as an implementation of this KMS interface. One of the things I'm trying to figure out is what, if anything, we need beyond that.

**Comment by HighWatersDev** [comment]
I use Sectigo. I am looking for an implementation of the ACME server to do exactly that: authenticate/authorize clients, validate requests, and then forward it to external CA authority, Sectigo in my case.

**Comment by mmalone** [comment]
@HighWatersDev I don't know much about Sectigo. I know they issue Web PKI certs for public websites, but I assume you're referring to their "Private PKI" product here? Do you know if they're using EJBCA or something under the hood, or do they have their own proprietary CA API that we'd need to integrate with?

Since Sectigo's Private PKI appears to be an enterprise offering we'd probably need to collaborate with them (or at least one of their customers) on an integration. Depending on their plans, they may not want us to deliver this in open source. If you're interested in working with us on this lemme know either here or via email (my email is mike at smallstep).

Honestly, though, Sectigo doesn't seem like low-hanging fruit for open source investment unless we hear more interest from other Sectigo users. It would be prioritized below a bunch of other integrations that are lower effort and where we have stronger interest: Vault, MS CA, GCP CAS, AWS ACM PCA, EJBCA, and cfssl (in rough order). That said, as soon as these integrations start rolling out there will be a plugin mechanism available and we'd _gladly_ accept additional contributed integrations. Also, if there's commercial interest here, that's a different conversation (one that we should probably move out of open source issues and into email :).

**Comment by rmhrisk** [comment]
@robstradling is Mr SectIgo and knows all the things but I do know that they have both an ACME API and a proprietary API.

What I am curious about @HighWatersDev is what is the use case, since they do have an ACME API how do you envision using smallstep with them?

**Comment by HighWatersDev** [comment]
Thanks for the insights!
@rmhrisk, my envision of using smallstep is pretty simple:
1. I have a fleet of nodes that are not accessible from the Internet.
2. I utilize both Let's Encrypt as well as Sectigo services.
3. The nodes will use some ACME client such as certbot (or smallstep in this case).
4. smallstep would come in between the fleet and PKI providers as an intermediary ACME server receiving requests from the nodes and forwarding them to LE/Sectigo, for example, using ACME protocol.
This is what I thought of when asking my question. Any feedback/suggestions are more than welcome!

**Comment by rmhrisk** [comment]
Makes sense

**Comment by maraino** [comment]
Support for a CA Service interface and CloudCAS has been [merged](https://github.com/smallstep/certificates/pull/367) to master. Now it should be possible to add support for signing using CAs and use step-ca as the "Registration Authority".

@dharanikumar-s @mmalone I'm going to mark this issue as closed as the support is now available The support for a specific implementation should be tracked in a different issue.

**Comment by mageru** [comment]
Not sure how implementation of two new protocols that indicate it should now be possible to add support for Vault indicate this issue (adding support for Vault) has been addressed in a way that closes the issue.

**Comment by dopey** [comment]
@mageru You're absolutely right. 

New issue to track this effort: https://github.com/smallstep/certificates/issues/642. 

Left a comment on the Vault issue as well: https://github.com/hashicorp/vault/issues/8690#issuecomment-880146382.

tldr, the plugin architecture now exists. If anyone would like to take a stab at writing the Vault plugin we'd be eager to assist.
